src/main/java/com/google/sps/data/LearningPathService.java

```diff
@@ -136,16 +137,64 @@ public LearningItem loadItem(long itemId) throws EntityNotFoundException {
 		return mapEntityToLearningItem(itemId, item);
 	}
 
-	private LearningItem mapEntityToLearningItem(long sectionId, Entity e) {
+    private LearningItem mapEntityToLearningItem(long sectionId, Entity e) {
 		return new LearningItem(e);
 	}
 
+    public ItemFeedback loadItemFeedback(long itemId) throws EntityNotFoundException {
+		Entity item = datastore.get(KeyFactory.createKey(ITEM_FEEDBACK, itemId));
+		return mapEntityToItemFeedback(itemId, item);
+	}
+    public List<ItemFeedback> loadItemFeedbacks(long learningPathId) throws EntityNotFoundException {
+		Query query = new Query(ITEM_FEEDBACK).addSort("id")
+				.setFilter(new Query.FilterPredicate("learningPath", Query.FilterOperator.EQUAL, learningPathId));
+
+		List<Entity> items = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+
+		return items.stream().map(e -> mapEntityToItemFeedback(learningPathId, e)).collect(Collectors.toList());
+	}
+
+    public List<ItemFeedback> loadItemFeedbacksUser(long userID) throws EntityNotFoundException {
+		Query query = new Query(ITEM_FEEDBACK).addSort("id")
+			.setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userID));
+
+		List<Entity> items = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+
+		return items.stream().map(e -> mapEntityToItemFeedback(userID, e)).collect(Collectors.toList());
+	}
+
+    private ItemFeedback mapEntityToItemFeedback(long sectionId, Entity e) {
+		return new ItemFeedback(e);
+	}
+
+    public double getSectionCompletion(long userId, LearningSection section ,long learningPath)throws EntityNotFoundException { 
+       // filter 
+        double feedbackCount = loadItemFeedbacksUser(userId).stream()
+                                    .filter( f -> f.getLearningSection() == section.getId() && f.isCompleted()).count();
+                                    
+        double proportion = feedbackCount/section.getNumItems();
+       
+        return proportion;
+    }
+
+    public LearningPath getLearningPathCompletion(long userId, long learningPathId)throws EntityNotFoundException {
+        LearningPath path = load(learningPathId);
+        List<LearningSection> sections = path.getSections();
+        long completion = 0;
+
+        for ( LearningSection s : sections){
+            completion += getSectionCompletion(userId, s, learningPathId);
```

Comment:
Note that this is computing an average. If sections have different lengths, this may produce interesting results:

 - Section 1: 1 item, done. "100%" complete
 - Section 2: 100 items, not done. "0% complete"
 - Path: 50% done (1/101 items).

That's fine for now, we didn't specify this in more detail.